arch/x86/kvm/pkvm/mmu.c

```diff
@@ -335,3 +392,105 @@ int pkvm_host_mmu_unmap(unsigned long vaddr, unsigned long size)
 	/* The vaddr == phys for the host MMU */
 	return pkvm_pgtable_unmap(&host_mmu, vaddr, vaddr, size);
 }
+
+/**
+ * pkvm_host_donate_hyp() - Donate memory pages from host to hypervisor.
+ * @phys:	Physical address of the memory region to donate.
+ * @size:	Size of the memory region to donate.
+ *
+ * The donation transfers ownership of the memory pages in range
+ * [@phys, @phys + @size) from the host to the hypervisor, thus protecting them
+ * from accessing by the host. The @phys and @size are required to be PAGE_SIZE
+ * aligned (@size is also required to be non-zeroed value).
+ *
+ * The donated memory pages are annotated with PKVM_ID_HYP owner id in the host
+ * mmu, and page states are updated to PKVM_PAGE_NONE, to indicate the ownership
+ * has been transferred to the hypervisor.
+ *
+ * Returns: 0 on success, or a negative error code on failure.
+ */
+int pkvm_host_donate_hyp(unsigned long phys, unsigned long size)
+{
+	int ret;
+
+	if (!PAGE_ALIGNED(phys) || !PAGE_ALIGNED(size) || size == 0)
+		return -EINVAL;
+
+	pkvm_host_mmu_lock();
+
+	ret = check_host_mem_pgstate(phys, size, PKVM_PAGE_OWNED);
+	if (ret)
+		goto unlock;
+
+	/* The vaddr == phys for the host MMU. */
+	ret = pkvm_pgtable_set_owner(&host_mmu, phys, size, PKVM_ID_HYP);
+	/*
+	 * pkvm_pgtable_set_owner() shouldn't fail here unless there is a bug.
+	 * Furthermore, if it fails, it means some (maybe not all) pages in the
+	 * range remain mapped in the host mmu, whereas their state will be
+	 * changed to PKVM_PAGE_NONE below, causing an inconsistency between the
+	 * page state and the mapping, which may lead to unexpected behavior. So
+	 * panic if it fails.
+	 */
+	BUG_ON(ret);
+
+	set_host_mem_pgstate(phys, size, PKVM_PAGE_NONE);
+unlock:
+	pkvm_host_mmu_unlock();
+
+	return ret;
+}
+
+/**
+ * pkvm_hyp_donate_host() - Donate memory pages from hypervisor to host.
+ * @phys:	Physical address of the memory region to donate.
+ * @size:	Size of the memory region to donate.
+ *
+ * The donation transfers ownership of the memory pages in range
+ * [@phys, @phys + @size) from the hypervisor to the host, thus allowing the
+ * host to access those pages. The @phys and @size are required to be PAGE_SIZE
+ * aligned (@size is also required to be non-zeroed value).
+ *
+ * The donated memory pages are mapped in the host mmu with read/write/exec
+ * permission and WB memory type in the host mmu, with the page states being
+ * updated to PKVM_PAGE_OWNED to indicate the ownership has been transferred to
+ * the host.
+ *
+ * Returns: 0 on success, or a negative error code on failure.
+ */
+int pkvm_hyp_donate_host(unsigned long phys, unsigned long size)
+{
+	u64 prot = host_mmu.pgt_ops->calc_pte_perm(true, true, true) |
+		   host_mmu.pgt_ops->calc_pte_memtype(false);
+	int ret;
+
+	if (!PAGE_ALIGNED(phys) || !PAGE_ALIGNED(size) || size == 0)
+		return -EINVAL;
```

Comment:
Late thought: would it make sense to make pkvm_hyp_donate_host() void and just BUG_ON in case of any failure in it, similarly to pkvm_host_unshare_hyp()? I see we never really check the return value of pkvm_hyp_donate_host()...